server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoGridAggregationBuilder.java

```diff
@@ -53,30 +51,76 @@
 
 public class GeoGridAggregationBuilder extends ValuesSourceAggregationBuilder<ValuesSource.GeoPoint, GeoGridAggregationBuilder>
         implements MultiBucketAggregationBuilder {
     public static final String NAME = "geohash_grid";
-    public static final int DEFAULT_PRECISION = 5;
     public static final int DEFAULT_MAX_NUM_CELLS = 10000;
 
     private static final ObjectParser<GeoGridAggregationBuilder, Void> PARSER;
     static {
         PARSER = new ObjectParser<>(GeoGridAggregationBuilder.NAME);
         ValuesSourceParserHelper.declareGeoFields(PARSER, false, false);
-        PARSER.declareField((parser, builder, context) -> builder.precision(parsePrecision(parser)), GeoHashGridParams.FIELD_PRECISION,
-            org.elasticsearch.common.xcontent.ObjectParser.ValueType.INT);
+        PARSER.declareString(GeoGridAggregationBuilder::type, GeoHashGridParams.FIELD_TYPE);
+        PARSER.declareField(GeoGridAggregationBuilder::parsePrecision, GeoHashGridParams.FIELD_PRECISION, ObjectParser.ValueType.INT);
         PARSER.declareInt(GeoGridAggregationBuilder::size, GeoHashGridParams.FIELD_SIZE);
         PARSER.declareInt(GeoGridAggregationBuilder::shardSize, GeoHashGridParams.FIELD_SHARD_SIZE);
     }
 
+    private static void parsePrecision(XContentParser parser, GeoGridAggregationBuilder builder, Void context)
+            throws IOException {
+        // In some cases, this value cannot be fully parsed until after we know the type
+        // Store precision as is until the end of parsing.
+        // ValueType.INT could be either a number or a string
+        builder.precisionLocation = parser.getTokenLocation();
+        if (parser.currentToken() == XContentParser.Token.VALUE_NUMBER) {
+            builder.precisionRaw = parser.intValue();
+        } else {
+            builder.precisionRaw = parser.text();
+        }
+    }
+
     public static GeoGridAggregationBuilder parse(String aggregationName, XContentParser parser) throws IOException {
-        return PARSER.parse(parser, new GeoGridAggregationBuilder(aggregationName), null);
+        GeoGridAggregationBuilder builder = PARSER.parse(parser,
+            new GeoGridAggregationBuilder(aggregationName), null);
+
+        try {
+            // delayed precision validation
+            final GeoHashTypeProvider typeHandler = builder.type.getHandler();
+
+            if (builder.precisionRaw == null) {
+                builder.precision(typeHandler.getDefaultPrecision());
+            } else if (builder.precisionRaw instanceof String) {
+                builder.precision(typeHandler.parsePrecisionString((String) builder.precisionRaw));
+            } else {
+                builder.precision((int) builder.precisionRaw);
+            }
+
+            return builder;
+        } catch (Exception e) {
+            throw new XContentParseException(builder.precisionLocation,
+                "[geohash_grid] failed to parse field [precision]", e);
+        }
```

Comment:
The precision stuff here looks messy to me. I think an alternative would be to pass the type into the constructor of the builder and make the type `final`. Then change the PARSER to be a `ConstructingObjectParser` so we can still parse the type but have the constructor use the result. Lastly we can then have the parsing of the precision work directly because it will always know the type by the time its called.